### PR TITLE
Feat/toast notifications logic

### LINKED ISSUE #7775: Display Toast Notifications
Display toast notification on top of the screen in reaction to user actions or system side effects.

### PR DESCRIPTION
**Pull request by PeKne**
## Description

Toast notifications are displayed according to user actions and system events.

## Related Issue

Resolve #7775 

## Screenshots:

https://user-images.githubusercontent.com/26143964/224655314-eda52b37-1269-402a-889e-bafb4dde7bb5.mov

### CODE REVIEW

File: suite-common/toast-notifications/src/notificationsActions.ts [hunk 1]
```diff
@@ -48,6 +54,13 @@ export const addToastOnce = createActionWithExtraDeps(
     },
 );
 
+export const removeToastByType = createAction(
+    `${ACTION_PREFIX}/removeToastByType`,
+    (payload: { toastType: ToastNotification['type'] }) => ({
```
Comment (juriczech) [inline suite-common/toast-notifications/src/notificationsActions.ts]: nit: consider creating `ToastNotificationType` type (sounds crazy in this sentence, but same thing as `NetworkSymbol`

File: suite-common/toast-notifications/src/types.ts [hunk 2]
```diff
@@ -158,9 +164,15 @@ export interface CommonNotificationPayload {
     error?: string;
 }
 
-export type ToastNotification = { context: 'toast' } & CommonNotificationPayload & ToastPayload;
-export type EventNotification = { context: 'event' } & CommonNotificationPayload &
-    NotificationEventPayload;
+type ToastContext = { context: 'toast' };
+type EventContext = { context: 'event' };
+
+export type ToastNotification = ToastContext & CommonNotificationPayload & ToastPayload;
```
Comment (juriczech) [inline suite-common/toast-notifications/src/types.ts]: nit: kind of makes sense to now rename it to `DesktopToastNotification` maybe?

File: suite-native/notifications/src/components/NotificationRenderer.tsx [hunk 3]
```diff
@@ -26,20 +30,28 @@ export const NotificationRenderer = ({ children }: NotificationRendererProps) =>
     const { applyStyle } = useNativeStyles();
     const { top: topSafeAreaInset } = useSafeAreaInsets();
     const transactionNotifications = useSelector(selectOpenedTransactionNotifications);
+    const toastNotifications = useSelector(selectNativeToastNotifications);
 
-    if (!transactionNotifications) return null;
+    if (!transactionNotifications && !toastNotifications) return null;
```
Comment (juriczech) [inline suite-native/notifications/src/components/NotificationRenderer.tsx]:
```suggestion
    if (!transactionNotifications || !toastNotifications) return null;
```

Comment (juriczech) [reply suite-native/notifications/src/components/NotificationRenderer.tsx]: You don't want to render it when either of them fails, right?

Comment (juriczech) [reply suite-native/notifications/src/components/NotificationRenderer.tsx]: Also shouldn't this be `A.isNotEmpty(transactionNotifications)`?

File: suite-native/notifications/src/components/ToastNotification.tsx [hunk 4]
```diff
@@ -46,54 +47,84 @@ const IconContainerStyle = prepareNativeStyle<{
...
+    'redirect-error': {
+        variant: 'error',
+        iconName: 'warningTriangle',
+        defaultMessage: 'Unable to open the link',
+    },
+} as const satisfies Record<
+    NativeToastType,
+    { variant: ToastNotificationVariant; iconName: IconName; defaultMessage: string }
+>;
+// todo change to toastNative tipe
```
Comment (juriczech) [inline suite-native/notifications/src/components/ToastNotification.tsx]: nit: I would consider creating constants file for this one as it might grow quite a bit in the future and doesn't need to be defined inside component. Resolve if you disagree

File: suite-native/notifications/src/components/ToastNotification.tsx [hunk 5]
```diff
@@ -46,54 +47,84 @@ const IconContainerStyle = prepareNativeStyle<{
...
         backgroundColor: 'backgroundAlertBlueSubtleOnElevation0',
         iconBackgroundColor: 'backgroundAlertBlueSubtleOnElevation1',
         textColor: 'textAlertBlue',
         iconColor: 'iconAlertBlue',
-        defaultIcon: 'info',
     },
 } as const satisfies Record<ToastNotificationVariant, ToastNotificationStyle>;
 
-export const ToastNotification = ({ variant = 'default', title, icon }: ToastNotificationProps) => {
+export const ToastNotification = memo(({ type, message }: ToastNotificationProps) => {
```
Comment (juriczech) [inline suite-native/notifications/src/components/ToastNotification.tsx]: Does this really need to be memoized or are you just using this "to be safe"?

**Review by juriczech — CHANGES_REQUESTED** [review]
Few nits, looks good 👌

File: suite-common/toast-notifications/src/notificationsSelectors.ts [hunk 6]
```diff
@@ -47,3 +49,13 @@ export const selectOpenedTransactionNotifications = (state: NotificationsRootSta
...
 };
+
+export const selectNativeToastNotifications = (state: NotificationsRootState) => {
+    const notifications = selectNotifications(state);
+
+    return notifications.filter(
+        n =>
+            n.context === 'toast' &&
+            nativeToastTypes.some(nativeToastType => nativeToastType === n.type),
+    ) as NativeToastNotification[];
```
Comment (juriczech) [inline suite-common/toast-notifications/src/notificationsSelectors.ts]: Please start using A.filter from @mobily/ts-belt
https://mobily.github.io/ts-belt/api/array/#filter

File: suite-native/notifications/src/components/NotificationRenderer.tsx (hunk 3 above)
Comment (PeKne) [reply suite-native/notifications/src/components/NotificationRenderer.tsx]: I need to fix this. It is completely wrong. If this expression would resolve to false, then all the content of the app would be hidden (oops, 😅), since it is rendered inside of the JSX as a child.

File: suite-common/toast-notifications/src/types.ts (hunk 2 above)
Comment (PeKne) [reply suite-common/toast-notifications/src/types.ts]: I would prefer to not touch anything that is used in the desktop version if you don't insist on this.

File: suite-native/notifications/src/components/ToastNotification.tsx (hunk 4 above)
Comment (PeKne) [reply suite-native/notifications/src/components/ToastNotification.tsx]: What about leaving it there and moving it away as it grows larger?

File: suite-native/notifications/src/components/ToastNotification.tsx (hunk 5 above)
Comment (PeKne) [reply suite-native/notifications/src/components/ToastNotification.tsx]: It does, if it's not memoized, then the timeout of hiding the toast is refreshed and takes longer to disappear then designed.

Comment (juriczech) [reply suite-native/notifications/src/components/ToastNotification.tsx]: Isn't that because you're missing `useEffect` deps?

Comment (PeKne) [reply suite-native/notifications/src/components/ToastNotification.tsx]: Oh, i missed that, you ar eprobably right. I will try to remove memo and add useEffect deps array.

File: suite-common/toast-notifications/src/notificationsActions.ts (hunk 1 above)
Comment (PeKne) [reply suite-common/toast-notifications/src/notificationsActions.ts]: fixed ce38fec3e

File: suite-native/notifications/src/components/NotificationRenderer.tsx (hunk 3 above)
Comment (PeKne) [reply suite-native/notifications/src/components/NotificationRenderer.tsx]: fixed ce38fec3e

File: suite-native/notifications/src/components/ToastNotification.tsx (hunk 5 above)
Comment (PeKne) [reply suite-native/notifications/src/components/ToastNotification.tsx]: fixed ce38fec3e

File: suite-common/toast-notifications/src/notificationsSelectors.ts (hunk 6 above)
Comment (PeKne) [reply suite-common/toast-notifications/src/notificationsSelectors.ts]: fixed ce38fec3e

File: suite-common/graph/src/hooks.ts [hunk 7]
```diff
@@ -58,6 +62,11 @@ export const useGraphForAccounts = ({
                 }
             } catch (err) {
                 setError(err?.message);
+                dispatch(
```
Comment (Nodonisko) [inline suite-common/graph/src/hooks.ts]: I don't know if we want to toast here, error is displayed in graph itself

File: suite-common/toast-notifications/src/notificationsActions.ts [hunk 8]
```diff
@@ -48,6 +54,13 @@ export const addToastOnce = createActionWithExtraDeps(
     },
 );
 
+export const removeToastByType = createAction(
```
Comment (Nodonisko) [inline suite-common/toast-notifications/src/notificationsActions.ts]: Name should be removeToast**s**ByType because it will remove all toasts by type.

Also why we need this?

File: suite-common/toast-notifications/src/notificationsSelectors.ts [hunk 9]
```diff
@@ -47,3 +49,13 @@ export const selectOpenedTransactionNotifications = (state: NotificationsRootSta
             (n.type === 'tx-received' || n.type === 'tx-sent' || n.type === 'tx-confirmed'),
     ) as TransactionNotification[];
 };
+
+export const selectNativeToastNotifications = (state: NotificationsRootState) => {
```
Comment (Nodonisko) [inline suite-common/toast-notifications/src/notificationsSelectors.ts]: In common only code related to both platforms should be present, otherwise move this to some native specific package.

File: suite-native/helpers/src/hooks/useRedirect.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import { Linking } from 'react-native';
+import { useDispatch } from 'react-redux';
+
+import { notificationsActions } from '@suite-common/toast-notifications';
+
+export const useRedirect = () => {
```
Comment (Nodonisko) [inline suite-native/helpers/src/hooks/useRedirect.ts]: please rename this to `useOpenLink` since there is nothing like redirect in mobile app

File: suite-native/helpers/src/hooks/useRedirect.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    const dispatch = useDispatch();
+
+    const handleRedirect = async (href: string) => {
```
Comment (Nodonisko) [inline suite-native/helpers/src/hooks/useRedirect.ts]: please wrap this to `useCallback`, in this universal hook you never know how people will use it and you should provide stable callback function signature

File: suite-native/module-dev-utils/src/screens/DemoScreen.tsx [hunk 12]
```diff
@@ -70,14 +61,6 @@ export const DemoScreen = () => {
                 <VStack>
                     <Text variant="titleSmall">Transaction Notifications:</Text>
                     <Text variant="titleSmall">Toast Notifications:</Text>
-                    <VStack
```
Comment (Nodonisko) [inline suite-native/module-dev-utils/src/screens/DemoScreen.tsx:81]: We won't have notifications in component demo? 😢

File: suite-native/module-settings/src/components/AboutUsBanners.tsx [hunk 13]
```diff
@@ -18,6 +19,7 @@ const stackStyle = prepareNativeStyle(_ => ({
 }));
 
 export const AboutUsBanners = () => {
+    const redirect = useRedirect();
```
Comment (Nodonisko) [inline suite-native/module-settings/src/components/AboutUsBanners.tsx]: openLink

File: suite-native/notifications/src/components/ToastNotification.tsx [hunk 14]
```diff
@@ -46,54 +47,84 @@ const IconContainerStyle = prepareNativeStyle<{
...
+        iconName: 'warningTriangle',
+        defaultMessage: 'Unable to open the link',
+    },
+} as const satisfies Record<
+    NativeToastType,
+    { variant: ToastNotificationVariant; iconName: IconName; defaultMessage: string }
+>;
+// todo change to toastNative tipe
+
+const ToastVariantToStyleMap = {
```
Comment (Nodonisko) [inline suite-native/notifications/src/components/ToastNotification.tsx]: no first letter capital

File: suite-native/notifications/src/components/ToastNotification.tsx [hunk 15]
```diff
@@ -46,54 +47,84 @@ const IconContainerStyle = prepareNativeStyle<{
     ),
 }));
 
-const ToastNotificationVariantToStyleMap = {
+const ToastTypeToContentMap = {
```
Comment (Nodonisko) [inline suite-native/notifications/src/components/ToastNotification.tsx]: no first letter capital

File: suite-native/notifications/src/components/ToastNotification.tsx [hunk 16]
```diff
@@ -46,54 +47,84 @@ const IconContainerStyle = prepareNativeStyle<{
...
+
+    useEffect(() => {
+        const timeout = setTimeout(
+            () => dispatch(notificationsActions.removeToastByType({ toastType: type })),
+            TOAST_VISIBLE_DURATION,
+        );
 
-    const iconName = icon ?? defaultIcon;
+        return () => clearTimeout(timeout);
+    });
```
Comment (Nodonisko) [inline suite-native/notifications/src/components/ToastNotification.tsx]: deps array is missing

File: suite-native/notifications/src/components/ToastNotification.tsx [hunk 17]
```diff
@@ -46,54 +47,84 @@ const IconContainerStyle = prepareNativeStyle<{
     ),
 }));
 
-const ToastNotificationVariantToStyleMap = {
+const ToastTypeToContentMap = {
+    'copy-to-clipboard': {
```
Comment (Nodonisko) [inline suite-native/notifications/src/components/ToastNotification.tsx]: Why do we need all these types? Why not just have type `toast` and pass all these info like `variant`, `iconName`, `message` with every call?

It will be much more flexible and we won't need to predefine all of possible types. Just do something like `dispatch(addNotification({ type: 'toast', icon: 'copy', variant: 'default', message: 'Hello copied.' })`

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/graph/src/hooks.ts (hunk 7 above)
Comment (PeKne) [reply suite-common/graph/src/hooks.ts]: Not sure actually. It was part of the issue assignment. Maybe @mnuky will know the answer?

File: suite-native/module-dev-utils/src/screens/DemoScreen.tsx (hunk 12 above)
Comment (PeKne) [reply suite-native/module-dev-utils/src/screens/DemoScreen.tsx:81]: No, since the data is read directly from the redux state. I could rework it a bit, but it would need some extra work. Do we really need it?

**Comment by PeKne** [comment]
@juriczech @Nodonisko I have reworked the toast notifications into the separate package `@suite-native/toast-notifications`. Thus many files changed since your last review and many of your comments are deprecated. Please review this PR again. 

FYI I also had to extract `Link` component and `useOpenLink` hook into a separate package `@suite-native/link` to avoid circular dependency between `atoms` and `toast-notifications` packages.

File: suite-native/link/src/components/Link.tsx [hunk 18]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import { TouchableOpacity, TouchableOpacityProps } from 'react-native';
+
+import { useOpenLink } from '../useOpenLink';
+
+type LinkProps = TouchableOpacityProps & {
```
Comment (Nodonisko) [inline suite-native/link/src/components/Link.tsx:6]: Please Omit `onPress` from props

File: suite-native/module-dev-utils/src/screens/DemoScreen.tsx (hunk 12 above)
Comment (Nodonisko) [reply suite-native/module-dev-utils/src/screens/DemoScreen.tsx:81]: Not really priority now

File: suite-native/toast-notifications/src/slice.ts [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+            { payload }: PayloadAction<ToastNotification>,
+        ) => void;
+    },
+    'toastNotifications'
+> = createSlice({
+    name: 'toastNotifications',
+    initialState: toastNotificationsInitialState,
+    reducers: {
+        addToastNotification: (state, { payload }: PayloadAction<ToastNotificationWithoutId>) => {
+            if (A.all(state, toastMessage => toastMessage.message !== payload.message)) {
```
Comment (Nodonisko) [inline suite-native/toast-notifications/src/slice.ts]: Do we need this check? I think it could make sense to have more toast with same text because for example if someone will click twice on something I would kind of expect that I will see toast twice as result of my actions.

File: suite-native/toast-notifications/src/slice.ts
Comment (Nodonisko) [inline suite-native/toast-notifications/src/slice.ts:1]: I am just wondering if we need redux for this. Basically I can replace this whole file with 10 lines of atoms:

```tsx
const notificationsAtom = atom<ToastNotification[]>([]);
const addNotificationAtom = atom(null, (get, set, notification) => {
    const notifications = get(notificationsAtom);
    set(notificationsAtom, [...notifications, notification]);
});
const removeNotificationAtom = atom(null, (get, set, id) => {
    const notifications = get(notificationsAtom);
    set(
        notificationsAtom,
        notifications.filter(n => n.id !== id),
    );
});
```

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-native/graph/src/components/Graph.tsx [hunk 20]
```diff
@@ -76,6 +78,20 @@ export const Graph = <TGraphPoint extends EnhancedGraphPoint>({
     onTryAgain,
     error,
 }: GraphProps<TGraphPoint>) => {
+    const dispatch = useDispatch();
+
+    useEffect(() => {
+        if (error) {
```
Comment (juriczech) [inline suite-native/graph/src/components/Graph.tsx]: nit: I prefer if not error return, since it avoids unnecessary nesting but it's up to you

File: suite-native/toast-notifications/src/slice.ts [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+            state: ToastNotificationsRootState,
+            { payload }: PayloadAction<ToastNotificationWithoutId>,
+        ) => void;
+        removeToastNotification: (
+            state: ToastNotificationsRootState,
+            { payload }: PayloadAction<ToastNotification>,
+        ) => void;
+    },
+    'toastNotifications'
+> = createSlice({
```
Comment (juriczech) [inline suite-native/toast-notifications/src/slice.ts]: this type shouldn't be necessary. Why did you add it?

**Review by juriczech — DISMISSED** [review]

File: suite-native/toast-notifications/src/slice.ts [hunk 22]
```diff
@@ -0,0 +1,48 @@
+import { createSlice, PayloadAction, Slice } from '@reduxjs/toolkit';
+import { A } from '@mobily/ts-belt';
+
+import {
+    ToastNotificationsRootState,
```
Comment (juriczech) [inline suite-native/toast-notifications/src/slice.ts]: Please define root state type in this file

File: suite-native/toast-notifications/src/slice.ts
Comment (PeKne) [reply suite-native/toast-notifications/src/slice.ts:1]: Do we have any clear rules to decide when to use redux/jotai? I have used redux since this state could have more selectors in future and I believe that for a more complex state is redux better.

But on the other hand, from the perspective of developer experience, I prefer Jotai. So I don't mind rewriting it.

File: suite-native/toast-notifications/src/slice.ts (hunk 19 above)
Comment (PeKne) [reply suite-native/toast-notifications/src/slice.ts]: From my point of view, it is logical to have this check. For example, if you would quickly click a "copy button" twenty times in a row when this check is missing, the screen would be flooded with toast notifications telling you that the content was copied to the clipboard. Do we really want this?

File: suite-native/toast-notifications/src/slice.ts (hunk 22 above)
Comment (PeKne) [reply suite-native/toast-notifications/src/slice.ts]: Ok, no problem. But why is that?

File: suite-native/toast-notifications/src/slice.ts (hunk 21 above)
Comment (PeKne) [reply suite-native/toast-notifications/src/slice.ts]: Copy-paste from the `module-settings` package. I will remove it there also.

File: suite-native/toast-notifications/src/slice.ts (hunk 22 above)
Comment (juriczech) [reply suite-native/toast-notifications/src/slice.ts]: just a convention across slices and reducers to have type in same file. Nothing too critical, same thing as you want to see type props in component

File: suite-native/toast-notifications/src/slice.ts
Comment (Nodonisko) [reply suite-native/toast-notifications/src/slice.ts:1]: I prefer to use jotai if possible and if there is no clear benefit of using redux. You can write selectors easily with jotai too and I don't expect this to be much more complicated that it is anyway so I vote for rewrite it.

File: suite-native/toast-notifications/src/slice.ts (hunk 19 above)
Comment (Nodonisko) [reply suite-native/toast-notifications/src/slice.ts]: Ok let's keep it for now. In future I would expect something like maximum number of toast that could be visible but for now it's good enough.

Comment (juriczech) [reply suite-native/toast-notifications/src/slice.ts]: Maybe a comment?

File: suite-native/graph/src/components/Graph.tsx (hunk 20 above)
Comment (Nodonisko) [reply suite-native/graph/src/components/Graph.tsx]: return in effect is for cleanup

Comment (juriczech) [reply suite-native/graph/src/components/Graph.tsx]: oh shit, it's useEffect. My bad 😅

File: suite-native/toast-notifications/src/slice.ts (hunk 22 above)
Comment (PeKne) [reply suite-native/toast-notifications/src/slice.ts]: deprecated since 148b2cb690ef72b5a2c4628595be1d22506c4b48

File: suite-native/toast-notifications/src/slice.ts (hunk 21 above)
Comment (PeKne) [reply suite-native/toast-notifications/src/slice.ts]: fixed in 148b2cb690ef72b5a2c4628595be1d22506c4b48

File: suite-native/toast-notifications/src/slice.ts
Comment (PeKne) [reply suite-native/toast-notifications/src/slice.ts:1]: state management converted to Jotai in 148b2cb690ef72b5a2c4628595be1d22506c4b48

File: suite-native/graph/src/components/Graph.tsx [hunk 23]
```diff
@@ -76,6 +78,16 @@ export const Graph = <TGraphPoint extends EnhancedGraphPoint>({
     onTryAgain,
     error,
 }: GraphProps<TGraphPoint>) => {
+    const addToastNotification = useSetAtom(addToastNotificationAtom);
```
Comment (Nodonisko) [inline suite-native/graph/src/components/Graph.tsx]: Please create hook that will wrap this so you don't need to import both `jotai` and `addToastNotificationAtom`. Something like `const { showToast } = useToast();` Also please used that named export through object for `showToast` so it must be named everywhere same.

File: suite-native/toast-notifications/package.json [hunk 24]
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "@suite-native/toast-notifications",
```
Comment (Nodonisko) [inline suite-native/toast-notifications/package.json]: Please drop `notifications` from all names everywhere. Just `toasts` is enough.

File: suite-native/toast-notifications/src/state.ts
Comment (Nodonisko) [inline suite-native/toast-notifications/src/state.ts:1]: please rename this file to `toastsAtoms.ts`

**Review by Nodonisko — CHANGES_REQUESTED** [review]

File: suite-common/graph/src/hooks.ts (hunk 7 above)
Comment (Nodonisko) [reply suite-common/graph/src/hooks.ts]: Let's remove it, error displayed in graph itself is enough.

File: suite-native/graph/src/components/Graph.tsx (hunk 23 above)
Comment (PeKne) [reply suite-native/graph/src/components/Graph.tsx]: 9c3195cce

File: suite-native/toast-notifications/package.json (hunk 24 above)
Comment (PeKne) [reply suite-native/toast-notifications/package.json]: 9c3195cce

File: suite-native/toast-notifications/src/state.ts
Comment (PeKne) [reply suite-native/toast-notifications/src/state.ts:1]: 9c3195cce

**Review by Nodonisko — APPROVED** [review]
